### DISCUSSION TITLE
Restoring Mnesia Database

### DISCUSSION THREAD

**Discussion by Arokan13** [processone/ejabberd#3895] [Questions] [answered]
Hey,

I'm using the docker-container on arm64, where before I used an old container v15.04.
I dumped the mnesia db to binary and made a backup in parallel and I'd like to restore the data.
Now, in order to do that, the server must be stopped which in turn stops the container, so I can't use ejabberdctl anymore.
Do you guys have an idea how to resolve this issue?

Also is there a way to export the mnesia db to another database as mnesia is limited to 2GB?

Thanks in advance!

**Comment by badlop** [comment]
I can think several ways to backup your mnesia database and later restore it in another machine:

A) Stop ejabberd, copy the [mnesia spool files](https://docs.ejabberd.im/admin/installation/#system-install) to the new location, then start the new ejabberd.

B) Use the [backup](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#backup) and [restore](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#restore) or [install](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#install-fallback) commands, which work with binary files. They require ejabberd be running.

C) Use the [dump](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#dump) and [load](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#load) commands, which work with text files, only suitable for small servers. Require ejabberd running

D) Use [export_piefxis](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#export-piefxis) and [import_piefxis](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#import-piefxis), which work with XML files, this may miss information as it supports only that XEP.

E) Use [export2sql](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#export2sql) command, and later import in an SQL database.

> I dumped the mnesia db to binary and made a backup in parallel and I'd like to restore the data.

Great!

> Now, in order to do that, the server must be stopped

No. If you used the backup, dump or export_piefxis commands, you are expected to start the new ejabberd node, and then run the corresponding command: restore/install, load, or import_piefxis


> Also is there a way to export the mnesia db to another database as mnesia is limited to 2GB?

Option E)

**Comment by jdolan-chwy** [comment]
Hi,

I have a similar question. I'm using an ejabberd Docker container from within unit tests which create various users, MUC rooms, etc. I'm hoping to take a `backup` of the database when the container first comes up and my admin account is created, and then `restore` that backup after each unit test runs. This will ensure that each test has a clean slate ejabberd instance to work with. However, this is not working as I'd assumed it would:


```
17:30:20.890 - DEBUG [main] EjabberdSupport : ejabberdctl backup /tmp/db -> Container.ExecResult(exitCode=0, stdout=, stderr=)

[ .. test runs here ]

17:30:46.702 - DEBUG [main] EjabberdSupport : ejabberdctl restore /tmp/db -> Container.ExecResult(exitCode=0, stdout=, stderr=)
17:30:47.595 - DEBUG [main] EjabberdSupport : ejabberdctl clear_cache -> Container.ExecResult(exitCode=0, stdout=, stderr=)
```

No errors, but on the very next test case that runs, it appears as tho the users and MUC rooms from the previous test are still there. Is there a way to purge ejabberd of data _without_ restarting the container between each test case?

**Reply by jdolan-chwy** [reply]
Follow-up, I've tried both backup / restore and dump / load. Basically the same behavior with both sets of commands.

**Reply by badlop** [reply, marked as answer]
> No errors, but on the very next test case that runs, it appears as tho the users and MUC rooms from the previous test are still there.

The [ejabberd function](https://github.com/processone/ejabberd/blob/6272c0e9011a6b866277f1ceee73d909a39cb61d/src/ejabberd_admin.erl#L825) that implements the `restore` API command explicitly keeps existing content, using the [keep_tables](https://www.erlang.org/doc/man/mnesia.html#restore-2) option when calling mnesia.

> Is there a way to purge ejabberd of data _without_ restarting the container between each test case?

Yes. For cases like yours, you may want to use [delete_mnesia](https://docs.ejabberd.im/developer/ejabberd-api/admin-api/#delete-mnesia) before restoring a backup.
